File: python/life/vegetarian.py
#vegetarian py

import random, math
from .config import *
from .genome import Genom, PERCENT_GENES
from .rules  import DEFAULT_RULES

class Vegetarian:

    # ── инициализация ────────────────────────────────────────────────────────
    def __init__(self, x=None, y=None, energy=None, genom=None, rules=None):
        # правила мира (цена статов, энергия растения, мутации); дети наследуют их
        self.rules = rules = rules or DEFAULT_RULES
        if genom is None:
            genom = VEGETARIAN_BASE_GENOM
        genom = Genom(*genom)          # принимаем и список (конфиг, тесты), и Genom

        self.genom = genom
        self.size  = genom.size
        self.speed = genom.speed
        self.vision = genom.vision
        self.repro_threshold = genom.repro_threshold
        self.repro_share     = genom.repro_share

        # ── гены слоя: диапазон высот (в процентах) ─────────────────────────
        min_pct = max(0, min(100, genom.min_y))   # clamp 0‒100
        max_pct = max(0, min(100, genom.max_y))
        if min_pct > max_pct:                     # гарантируем min ≤ max
            min_pct, max_pct = max_pct, min_pct

        self.min_y = min_pct      # храним в процентах, т.к. так уже использовали
        self.max_y = max_pct

        # ── границы ─────────────────────────────────────────────────────────
        # Границы слоя без запаса на тело — по ним решается, дотянемся ли до еды;
        # с запасом — полоса, в которой держится само тело, чтобы не торчало.
        self.layer_lo = (min_pct / 100) * WORLD_HEIGHT
        self.layer_hi = (max_pct / 100) * WORLD_HEIGHT
        # Запас на тело — не больше половины мира. Размер — ген, и при дешёвом
        # размере (лаборатория) тело бывает больше мира: с полным запасом границы
        # переворачивались (x_lo > x_hi), зажимы в move() перекидывали существо
        # от края к краю на сотни пикселей и выталкивали за мир. Тело шире мира
        # целиком не уместить — такое существо держится на средней линии.
        margin_x = min(self.size, WORLD_WIDTH / 2)
        margin_y = min(self.size, WORLD_HEIGHT / 2)
        body_lo = self.layer_lo + margin_y
        body_hi = self.layer_hi - margin_y
        # Слой может оказаться уже собственного тела (эволюция сводит min_y и max_y),
        # и полоса переворачивается. Схлопываем её в линию посередине — один раз и
        # для всех: и для рождения, и для move(), и для выбора цели. Середину держим
        # в мире: у слоя на самом краю она легла бы за границу.
        if body_lo > body_hi:
            body_lo = body_hi = min(max((body_lo + body_hi) / 2, margin_y),
                                    WORLD_HEIGHT - margin_y)
        self.body_lo, self.body_hi = body_lo, body_hi
        self.x_lo = margin_x
        self.x_hi = WORLD_WIDTH - margin_x

        # ─────────────────────────────────────────────────────────────────────
        self.max_energy = self.size * VEGETARIAN_ENERGY_PER_SIZE
        # не переливать: у ребёнка свой бак, и он может быть меньше родительского
        self.energy = (min(energy, self.max_energy) if energy is not None
                       else self.max_energy * 0.5)
        self.alive  = True

        self.flee_ticks = 0
        self.flee_dx = self.flee_dy = 0.0   # последний вектор бегства (единичный)

        # Позиция всегда внутри своей полосы — и случайная, и заданная. Ребёнок
        # рождается у родителя, а слой у него уже свой, мутировавший: без зажима
        # первый же ход телепортировал бы его в слой, бывало на 2000 px.
        # uniform, а не randint: размер — ген, он дробный, а randint дробных не берёт.
        if x is None:
            x = random.uniform(self.x_lo, self.x_hi)
        if y is None:
            y = random.uniform(body_lo, body_hi)
        self.x = min(max(x, self.x_lo), self.x_hi)
        self.y = min(max(y, body_lo), body_hi)

        # ── предвычисленное ──────────────────────────────────────────────────
        # Геном не меняется всю жизнь, поэтому всё производное от него считается
        # один раз здесь, а не заново на каждом тике в move(). Особенно расход
        # энергии: три возведения в степень за тик на каждое существо.
        self.upkeep = rules.upkeep(self.size, self.speed, self.vision)
        # квадраты радиусов: сравнивать квадраты расстояний дешевле, чем звать hypot
        self.vision2 = self.vision * self.vision
        self.size2   = self.size * self.size
        self.half    = self.size / 2        # радиус тела: по нему ловят хищники
        self.flee2   = (self.vision / 3) * (self.vision / 3)

    # Функция движения
    def _pick_random_target(self):
        """Случайная точка в пределах vision и своего вертикального слоя."""
        lo, hi = self.body_lo, self.body_hi
        x, y, vision = self.x, self.y, self.vision
        x_lo, x_hi = self.x_lo, self.x_hi
        # слой схлопнут в линию: случайная точка на неё не попадёт никогда,
        # и существо стояло бы столбом — гуляем только вдоль линии
        flat = lo == hi

        for _ in range(10):
            angle = random.uniform(0, 2 * math.pi)
            dist  = random.uniform(vision * 0.5, vision * 2)
            tx = x + math.cos(angle) * dist
            ty = lo if flat else y + math.sin(angle) * dist
            if x_lo <= tx <= x_hi and lo <= ty <= hi:
                self.tx, self.ty = tx, ty
                return
        self.tx, self.ty = x, y  # fallback

    # move() — самое горячее место всей симуляции: он вызывается за каждое
    # травоядное каждый тик. Отсюда локальные переменные вместо self.x в циклах,
    # квадраты расстояний вместо hypot и развёрнутые вручную min(max(...)).
    def move(self, plants, predators):
        x, y, speed = self.x, self.y, self.speed

        # ── ищем ближайшего хищника в пределах vision ───────────────────────
        nearest_pred, best_pred_d2 = None, self.vision2
        for pr in predators:
            dx = pr.x - x
            dy = pr.y - y
            d2 = dx * dx + dy * dy
            if d2 < best_pred_d2:
                nearest_pred, best_pred_d2 = pr, d2

        fleeing = False
        if self.flee_ticks > 0:
            fleeing = True
            self.flee_ticks -= 1
        if nearest_pred is not None and best_pred_d2 < self.flee2:
            self.flee_ticks = FLEE_TICKS
            fleeing = True

        if fleeing:
            # пока хищник виден — обновляем направление бегства;
            # когда он пропал из виду, продолжаем бежать по последнему вектору
            if nearest_pred is not None:
                dx, dy = x - nearest_pred.x, y - nearest_pred.y
                d = math.hypot(dx, dy)
                if d != 0:
                    self.flee_dx, self.flee_dy = dx / d, dy / d

            tx = x + self.flee_dx * speed
            ty = y + self.flee_dy * speed
        else:
            # поиск ближайшего растения
            nearest, best_d2 = None, self.vision2
            for p in plants:
                if not p.alive:            # съедено раньше в этом же тике
                    continue
                dx = p.x - x
                dy = p.y - y
                d2 = dx * dx + dy * dy
                if d2 < best_d2:
                    nearest, best_d2 = p, d2

            # проверяем попадает ли еда в наш вертикальный слой.
            # ВАЖНО: сперва ищем ближайшее вообще и только потом отбрасываем.
            # Слить проверку слоя внутрь цикла — значит искать ближайшее В СЛОЕ,
            # а это уже другое поведение: существо перестанет отвлекаться на еду,
            # до которой не дотянуться, и пойдёт к следующей вместо блуждания.
            if nearest is not None and not (self.layer_lo <= nearest.y <= self.layer_hi):
                nearest = None

            if nearest is not None:
                tx, ty = nearest.x, nearest.y
            else:
                # hasattr, а не tx=x в __init__: существо, родившееся во время
                # бегства, иначе пошло бы потом к месту своего рождения
                if not hasattr(self, "tx"):
                    self._pick_random_target()
                else:
                    dx, dy = x - self.tx, y - self.ty
                    if dx * dx + dy * dy < speed * speed:
                        self._pick_random_target()
                tx, ty = self.tx, self.ty

        # вектор и шаг
        dx, dy = tx - x, ty - y
        d = math.hypot(dx, dy)
        if d != 0:
            step = speed / d      # именно так, а не (dx*speed)/d — округление разное
            dx *= step
            dy *= step

        # Новое положение: по x — границы мира, по y — своя полоса. Полоса
        # (см. __init__) всегда лежит внутри мира, поэтому отдельный зажим по
        # высоте мира не нужен. Существо уже стоит внутри своих границ, так что
        # зажим только укорачивает шаг: за тик оно сдвигается не дальше speed.
        nx = x + dx
        if   nx < self.x_lo: nx = self.x_lo
        elif nx > self.x_hi: nx = self.x_hi

        ny = y + dy
        if   ny < self.body_lo: ny = self.body_lo
        elif ny > self.body_hi: ny = self.body_hi

        self.x, self.y = nx, ny

        # энергозатраты (посчитаны один раз в __init__ — геном не меняется)
        self.energy -= self.upkeep
        if self.energy <= 0:
            self.alive = False

    def try_eat(self, plants: list) -> bool:
        """
        Съесть все растения не дальше `self.size` от центра.
        Возвращает True, если хоть что-то съели.

        size — диаметр тела (так оно и рисуется), а дотягивается существо на
        целый size, то есть на полтела дальше своего края.

        `plants` — кандидаты от сетки соседей, а не весь мир (см. world.py).
        Съеденное помечается alive=False, а не вырезается из списка: вырезание
        стоило линейного поиска на каждое растение и сломало бы кэш сетки.
        Мёртвые выметаются один раз за тик в World.
        """
        eaten = 0
        r2 = self.size2                     # квадрат радиуса поедания
        cx, cy = self.x, self.y

        for p in plants:
            if not p.alive:
                continue
            dx = cx - p.x
            dy = cy - p.y
            if dx * dx + dy * dy <= r2:
                p.alive = False
                eaten += 1

        if not eaten:
            return False

        # пополняем энергию
        gain = self.rules.plant_energy * eaten
        self.energy = min(self.max_energy, self.energy + gain)

        # ► сразу берём новую случайную цель — перестаём топтаться на месте
        self._pick_random_target()

        return True

    def mutate(self, sigma=None):
        if sigma is None:
            sigma = self.rules.mutation_sigma
        new_genom = []
        for name, value in zip(Genom._fields, self.genom):
            while True:
                gauss = random.gauss(0, sigma)
                if gauss >= -0.9:
                    break
            mutated = value * (1 + gauss)
            # гены-проценты (порог, доля потомку, слой) держим в 0‒100
            if name in PERCENT_GENES:
                mutated = min(100, max(0, mutated))
            new_genom.append(max(0.01, mutated))
        return Genom(*new_genom)

    def maybe_divide(self, offspring: list):
        """Размножаемся, если после деления у родителя остаётся резерв.

        Детей кладём в отдельный список: в этом тике они не ходят.
        """
        threshold = self.max_energy * (self.repro_threshold / 100)
        if self.energy < threshold + VEGETARIAN_REPRO_RESERVE:
            return                      # энергии недостаточно

        # Резерв проверяется ещё и ПОСЛЕ дележа: доля ребёнка считается от всей
        # энергии, и без этой проверки родитель, бывало, отдавал всё до нуля и
        # ниже — и умирал на следующем ходу, а хищник, съевший такого, терял энергию.
        child_energy = self.energy * (self.repro_share / 100)
        left = self.energy - child_energy - VEGETARIAN_REPRO_COST
        if left < VEGETARIAN_REPRO_RESERVE:
            return

        self.energy = left                                  # родитель платит
        child_genom = self.mutate()

        # Смещения по осям независимые: с одним общим дети ложились строго на
        # диагональ от родителя. В мир и в слой ребёнка зажимает его __init__.
        span = self.size * 2
        cx = self.x + random.uniform(-span, span)
        cy = self.y + random.uniform(-span, span)

        offspring.append(
            Vegetarian(x=cx, y=cy, energy=child_energy, genom=child_genom,
                       rules=self.rules)
        )

File: crates/life-core/src/vegetarian.rs
//! Травоядное: ищет растения в своём слое глубины, бежит от хищников, делится.
//!
//! Поиск соседей — забота мира. Существо получает его в виде замыканий
//! «найди ближайшее растение / ближайшего хищника»: так оно не знает о сетке,
//! а тесты подсовывают вместо неё обычный список. Замыкания ленивые: пока
//! существо бежит, растения оно не ищет вовсе.

use crate::config::*;
use crate::genome::Genom;
use crate::rng::Rng;
use crate::rules::Rules;
use crate::space::Space;

#[derive(Clone, Debug)]
pub struct Vegetarian {
    /// Постоянный номер: по нему игра выбирает и следит за существом.
    pub id: u64,
    pub x: f64,
    pub y: f64,
    pub energy: f64,
    pub alive: bool,

    pub genom: Genom,
    pub size: f64,
    pub speed: f64,
    pub vision: f64,

    // ── слой обитания и границы (из генома, считаются при рождении) ─────────
    /// Слой без запаса на тело — по нему решается, дотянемся ли до еды.
    pub layer_lo: f64,
    pub layer_hi: f64,
    /// Полоса, в которой держится само тело. Всегда внутри мира.
    pub body_lo: f64,
    pub body_hi: f64,
    pub x_lo: f64,
    pub x_hi: f64,

    // ── предвычисленное: геном не меняется всю жизнь ────────────────────────
    pub max_energy: f64,
    pub upkeep: f64,
    pub vision2: f64,
    pub size2: f64,
    /// Радиус тела: по нему ловят и видят хищники.
    pub half: f64,
    flee2: f64,

    pub flee_ticks: u32,
    /// Последний вектор бегства (единичный): бежим по нему, и когда хищник
    /// пропал из виду.
    flee_dx: f64,
    flee_dy: f64,
    /// Цель блуждания. None до первого выбора: существо, родившееся во время
    /// бегства, иначе пошло бы потом к месту своего рождения.
    target: Option<(f64, f64)>,
    pub rng: Rng,
}

impl Vegetarian {
    /// Новое травоядное. Координата None — случайная в своей полосе; заданная
    /// тоже зажимается в полосу: ребёнок рождается у родителя, а слой у него
    /// свой, мутировавший, и без зажима первый ход телепортировал бы его.
    /// energy None — полбака; заданная не больше бака.
    pub fn new(
        space: &Space,
        rules: &Rules,
        genom: Genom,
        x: Option<f64>,
        y: Option<f64>,
        energy: Option<f64>,
        mut rng: Rng,
    ) -> Self {
        let size = genom.size;
        let (mut min_pct, mut max_pct) = (genom.min_y.clamp(0.0, 100.0), genom.max_y.clamp(0.0, 100.0));
        if min_pct > max_pct {
            std::mem::swap(&mut min_pct, &mut max_pct);
        }
        let layer_lo = min_pct / 100.0 * space.height;
        let layer_hi = max_pct / 100.0 * space.height;

        // Запас на тело — не больше половины мира. Размер — ген, и при дешёвом
        // размере (лаборатория) тело бывает больше мира: с полным запасом
        // границы переворачивались, и зажимы перекидывали существо от края к краю.
        let margin_x = size.min(space.width / 2.0);
        let margin_y = size.min(space.height / 2.0);
        let (mut body_lo, mut body_hi) = (layer_lo + margin_y, layer_hi - margin_y);
        // Слой уже собственного тела — схлопываем полосу в линию посередине,
        // держа её в мире: у слоя на самом краю середина легла бы за границу.
        if body_lo > body_hi {
            let mid = ((body_lo + body_hi) / 2.0).clamp(margin_y, space.height - margin_y);
            body_lo = mid;
            body_hi = mid;
        }
        let (x_lo, x_hi) = (margin_x, space.width - margin_x);

        let max_energy = size * VEGETARIAN_ENERGY_PER_SIZE;
        let energy = match energy {
            Some(e) => e.min(max_energy),
            None => max_energy * 0.5,
        };

        let x = x.unwrap_or_else(|| rng.uniform(x_lo, x_hi)).clamp(x_lo, x_hi);
        let y = y.unwrap_or_else(|| rng.uniform(body_lo, body_hi)).clamp(body_lo, body_hi);
        let vision = genom.vision;

        Vegetarian {
            id: 0,
            x,
            y,
            energy,
            alive: true,
            genom,
            size,
            speed: genom.speed,
            vision,
            layer_lo,
            layer_hi,
            body_lo,
            body_hi,
            x_lo,
            x_hi,
            max_energy,
            upkeep: rules.upkeep(size, genom.speed, vision),
            vision2: vision * vision,
            size2: size * size,
            half: size / 2.0,
            flee2: (vision / 3.0) * (vision / 3.0),
            flee_ticks: 0,
            flee_dx: 0.0,
            flee_dy: 0.0,
            target: None,
            rng,
        }
    }

    /// Базовое травоядное из конфига в случайном месте.
    pub fn base(space: &Space, rules: &Rules, rng: Rng) -> Self {
        Vegetarian::new(space, rules, Genom::from_array(VEGETARIAN_BASE_GENOM), None, None, None, rng)
    }

    /// Случайная точка в пределах зрения и своей полосы; 10 попыток, иначе стоим.
    fn pick_random_target(&mut self) {
        let (lo, hi) = (self.body_lo, self.body_hi);
        // слой схлопнут в линию: случайная точка на неё не попадёт никогда,
        // и существо стояло бы столбом — гуляем только вдоль линии
        let flat = lo == hi;
        for _ in 0..10 {
            let angle = self.rng.uniform(0.0, std::f64::consts::TAU);
            let dist = self.rng.uniform(self.vision * 0.5, self.vision * 2.0);
            let tx = self.x + angle.cos() * dist;
            let ty = if flat { lo } else { self.y + angle.sin() * dist };
            if self.x_lo <= tx && tx <= self.x_hi && lo <= ty && ty <= hi {
                self.target = Some((tx, ty));
                return;
            }
        }
        self.target = Some((self.x, self.y));
    }

    /// Один ход.
    ///
    /// `nearest_predator(x, y, r2)` — ближайший хищник строго ближе √r2, или None.
    /// `nearest_plant(x, y, r2)` — ближайшее живое растение строго ближе √r2.
    pub fn step(
        &mut self,
        nearest_predator: impl FnOnce(f64, f64, f64) -> Option<(f64, f64, f64)>,
        nearest_plant: impl FnOnce(f64, f64, f64) -> Option<(f64, f64)>,
    ) {
        let (x, y, speed) = (self.x, self.y, self.speed);

        // (x, y, квадрат расстояния) ближайшего хищника в пределах зрения
        let pred = nearest_predator(x, y, self.vision2);

        let mut fleeing = false;
        if self.flee_ticks > 0 {
            fleeing = true;
            self.flee_ticks -= 1;
        }
        if let Some((_, _, d2)) = pred
            && d2 < self.flee2
        {
            self.flee_ticks = FLEE_TICKS;
            fleeing = true;
        }

        let (tx, ty) = if fleeing {
            if let Some((px, py, _)) = pred {
                let (dx, dy) = (x - px, y - py);
                let d = dx.hypot(dy);
                if d != 0.0 {
                    self.flee_dx = dx / d;
                    self.flee_dy = dy / d;
                }
            }
            (x + self.flee_dx * speed, y + self.flee_dy * speed)
        } else {
            // ВАЖНО: сперва ищем ближайшее вообще и только потом отбрасываем
            // чужой слой. Искать ближайшее В СЛОЕ — другое поведение: существо
            // перестанет отвлекаться на недосягаемую еду и не будет блуждать.
            let food = nearest_plant(x, y, self.vision2)
                .filter(|&(_, py)| self.layer_lo <= py && py <= self.layer_hi);
            match food {
                Some(p) => p,
                None => {
                    match self.target {
                        None => self.pick_random_target(),
                        Some((tx, ty)) => {
                            let (dx, dy) = (x - tx, y - ty);
                            if dx * dx + dy * dy < speed * speed {
                                self.pick_random_target();
                            }
                        }
                    }
                    self.target.unwrap()
                }
            }
        };

        let (mut dx, mut dy) = (tx - x, ty - y);
        let d = dx.hypot(dy);
        if d != 0.0 {
            let k = speed / d;
            dx *= k;
            dy *= k;
        }
        // Существо уже стоит внутри своих границ, так что зажим только укорачивает
        // шаг: за ход оно сдвигается не дальше speed.
        self.x = (x + dx).clamp(self.x_lo, self.x_hi);
        self.y = (y + dy).clamp(self.body_lo, self.body_hi);

        self.energy -= self.upkeep;
        if self.energy <= 0.0 {
            self.alive = false;
        }
    }

    /// Правила поменялись посреди жизни (лаборатория на ходу): пересчитать то,
    /// что при рождении было вычислено из правил. Как в `new`.
    pub fn apply_rules(&mut self, rules: &Rules) {
        self.upkeep = rules.upkeep(self.size, self.speed, self.vision);
    }

    /// Съедено `eaten` растений: энергия и сразу новая цель, чтобы не топтаться.
    pub fn feed(&mut self, eaten: usize, rules: &Rules) {
        if eaten == 0 {
            return;
        }
        self.energy = self.max_energy.min(self.energy + rules.plant_energy * eaten as f64);
        self.pick_random_target();
    }

    /// Ребёнок, если после деления у родителя остаётся резерв. Номер ребёнку
    /// выдаёт мир.
    pub fn maybe_divide(&mut self, space: &Space, rules: &Rules) -> Option<Vegetarian> {
        let threshold = self.max_energy * (self.genom.repro_threshold / 100.0);
        if self.energy < threshold + VEGETARIAN_REPRO_RESERVE {
            return None;
        }
        // Резерв проверяется и ПОСЛЕ дележа: доля ребёнка считается от всей
        // энергии, и без этого родитель отдавал всё до нуля и умирал.
        let child_energy = self.energy * (self.genom.repro_share / 100.0);
        let left = self.energy - child_energy - VEGETARIAN_REPRO_COST;
        if left < VEGETARIAN_REPRO_RESERVE {
            return None;
        }
        self.energy = left;
        let genom = self.genom.mutate(rules.mutation_sigma, &mut self.rng);

        // Смещения по осям независимые: с одним общим дети ложились на диагональ.
        let span = self.size * 2.0;
        let cx = self.x + self.rng.uniform(-span, span);
        let cy = self.y + self.rng.uniform(-span, span);
        let rng = self.rng.fork();
        Some(Vegetarian::new(space, rules, genom, Some(cx), Some(cy), Some(child_energy), rng))
    }
}
